src/btrfs.c

```diff
@@ -0,0 +1,183 @@
+/**
+ * collectd - src/btrfs.c
+ * Copyright (C) 2018       Martin Langlotz
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; only version 2 of the License is applicable.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
+ *
+ * Authors:
+ *   Martin Langlotz < stackshadow at evilbrain . de >
+ **/
+
+#include "collectd.h"
+
+#include "common.h"
+#include "plugin.h"
+#include "utils_avltree.h"
+#include "utils_ignorelist.h"
+#include "utils_mount.h"
+
+#include <btrfs/ioctl.h>
+
+#define PLUGIN_NAME "btrfs"
+static const char *config_keys[] = {"Path"};
+static int config_keys_num = STATIC_ARRAY_SIZE(config_keys);
+static char *conf_btrfs_path;
+static c_avl_tree_t *conf_list_btrfs_paths = NULL;
+
+static int btrfs_avl_compare(const void *key, const void *value) { return 1; };
+
+static int btrfs_config(const char *key, const char *value) {
+
+  // create linked-list
+  if (conf_list_btrfs_paths == NULL) {
+    conf_list_btrfs_paths = c_avl_create(btrfs_avl_compare);
```

Comment:
Check for `c_avl_create` success is missing here.

Also, personally I don't like then things are used not as they intended for.
Linked list (from `utils_llist.h`) should be used for a such task.
Shutdown callback should be implemented for a good form too.
Please look into tail of `src/powerdns.c` for examples.

Reply:
Yeah, thats right, it looks very creepy. I used it, because there was an iterator-function, which i missed at llist. After i look on powerdns.c its clear how llist work, so yes, i will rework that :)

thanks for that hint :+1: